src/main/java/org/openpnp/machine/reference/vision/ReferenceBottomVision.java

```diff
@@ -65,10 +68,77 @@ public PartAlignmentOffset findOffsets(Part part, BoardLocation boardLocation,
+			double placementAngle=angle;
+            MovableUtils.moveToLocationAtSafeZ(nozzle, VisionUtils.getBottomVisionCamera().getLocation().add(new Location(part.getHeight().getUnits(), 0.0, 0.0,part.getHeight().getValue(), 0.0)).derive(null, null, null, angle));
+            CvPipeline pipeline = partSettings.getPipeline();
+            pipeline.setCamera(VisionUtils.getBottomVisionCamera());
+            pipeline.setNozzle(nozzle);
+			pipeline.setValue(angle);
+			if(Boolean.getBoolean("develop")) {
+            if ("Rotate".equals(pipeline.getStages().get(1).getClass().getSimpleName())) {
+                ((Rotate)pipeline.getStages().get(1)).setDegrees( angle );
+            }}
+			pipeline.process();
+			if (!((pipeline.getResult("result")).model instanceof RotatedRect)) {
+				throw new Exception("Bottom vision alignment failed for part " + part.getId()
+						+ " on nozzle " + nozzle.getName() + ". No result found.");
+			}
+
+
+			RotatedRect rect = ((RotatedRect)(pipeline.getResult("result")).model);
+			if (rect.size.width < rect.size.height) { rect.angle = 90 + rect.angle; }
+            Logger.debug("Result rect {}", rect);
+			while (Math.abs(angle) > 45.0) { nozzleAngle= (angle += (angle < 0.)? 90 : -90); } 
+			while (Math.abs(rect.angle) > 45.0) { rect.angle += (rect.angle < 0)? 90 : -90; } 
+			angle += rect.angle;
+			while (Math.abs(angle) > 45.0) { angle += (angle < 0.)? 90 : -90; } 
+
+			int cnt=1;
+			if (pipeline.getResult("contours")!=null&&pipeline.getResult("contours").model instanceof List<?>) {
+				cnt=((List<?>)(pipeline.getResult("contours")).model).size();
+			}
+
+			String s="";
+			Location offsets = null;
+			int old=999;
+			int rot=999; do {	
+				offsets = new Location(LengthUnit.Millimeters,Double.NaN, Double.NaN, Double.NaN, placementAngle+angle);
+				nozzle.moveTo(offsets, part.getSpeed());
+				pipeline.process();
+				if (!((pipeline.getResult("result")).model instanceof RotatedRect)) {
+					throw new Exception("Bottom vision alignment failed for part " + part.getId()
+							+ " on nozzle " + nozzle.getName() + ". No result found.");
+				}
+
+				rect = (RotatedRect) pipeline.getResult("result").model;
+		                Logger.debug("Result rect {} inside vision loop", rect);
+				offsets = VisionUtils.getPixelCenterOffsets(pipeline.getCamera(), rect.center.x, rect.center.y).derive(null,null,null,-angle);
+				s = "Align offset for "+part.getName()+": " + offsets.toString() + "     " ;
+				MainFrame.get().setStatus(s);
+				offsets = offsets.derive(null,null,null,Double.NaN);
+				if(cnt>Integer.getInteger("contours",40)||Double.isNaN(pipeline.getValue())) {
+					if (rect.size.width < rect.size.height) { rect.angle = 90 + rect.angle; }
+					while (Math.abs(rect.angle) > 45.0) { rect.angle += (rect.angle < 0)? 90 : -90; } 
+					angle=nozzleAngle + rect.angle;
+					while (Math.abs(angle) > 45.0) { angle += (angle < 0.)? 90 : -90; }
+					double step=(360./200./16.);
```

Comment:
Why 360/200/16? That looks like it's assuming 16 microsteps on a direct drive nozzle. Is that important? Will it work with other types of nozzles?